File: android/app/src/main/java/com/rexvane/inkhole/p2p/DeviceIdentity.kt
package com.rexvane.inkhole.p2p

import android.content.Context
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.util.Base64
import java.io.ByteArrayOutputStream
import java.io.DataOutputStream
import java.math.BigInteger
import java.security.AlgorithmParameters
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.KeyStore
import java.security.MessageDigest
import java.security.Signature
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec
import java.security.spec.ECParameterSpec
import java.security.spec.ECPoint
import java.security.spec.ECPublicKeySpec

/** Hardware-backed when available; private key material never leaves Android Keystore. */
class DeviceIdentity(@Suppress("UNUSED_PARAMETER") context: Context) {
    companion object {
        private const val ALIAS = "inkhole-lan-identity-v1"
    }

    private val store = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }

    init {
        if (!store.containsAlias(ALIAS)) {
            KeyPairGenerator.getInstance(
                KeyProperties.KEY_ALGORITHM_EC, "AndroidKeyStore").apply {
                initialize(KeyGenParameterSpec.Builder(
                    ALIAS,
                    KeyProperties.PURPOSE_SIGN or KeyProperties.PURPOSE_VERIFY,
                ).setAlgorithmParameterSpec(ECGenParameterSpec("secp256r1"))
                    .setDigests(KeyProperties.DIGEST_SHA256)
                    .build())
                generateKeyPair()
            }
        }
    }

    private val entry: KeyStore.PrivateKeyEntry
        get() = store.getEntry(ALIAS, null) as KeyStore.PrivateKeyEntry

    val publicBytes: ByteArray
        get() = DeviceAuth.encodePublicKey(entry.certificate.publicKey as ECPublicKey)

    val publicKey: String
        get() = Base64.encodeToString(publicBytes, Base64.NO_WRAP)

    val fingerprint: String
        get() = DeviceAuth.hex(MessageDigest.getInstance("SHA-256").digest(publicBytes))

    fun sign(message: ByteArray): String {
        val signature = Signature.getInstance("SHA256withECDSA").apply {
            initSign(entry.privateKey)
            update(message)
        }.sign()
        return Base64.encodeToString(signature, Base64.NO_WRAP)
    }
}

object DeviceAuth {
    private val CAP_DOMAIN = "INKHOLE-WHPC3\u0000".toByteArray(Charsets.US_ASCII)
    private val TRANSFER_DOMAIN = "INKHOLE-WHPP3-AUTH\u0000".toByteArray(Charsets.US_ASCII)
    private val RECEIVER_DOMAIN =
        "INKHOLE-WHPP3-RECEIVER\u0000".toByteArray(Charsets.US_ASCII)

    fun capabilityMessage(nonce: ByteArray, instanceId: String, peerName: String,
                          version: Int, capabilities: Collection<String>): ByteArray {
        require(nonce.size == 32)
        require(version in 0..0xffff)
        val name = peerName.toByteArray(Charsets.UTF_8)
        require(name.size <= 0xffff)
        val caps = capabilities.toSortedSet().map { value ->
            value.toByteArray(Charsets.UTF_8).also { require(it.size <= 0xffff) }
        }
        require(caps.size <= 0xffff)
        return ByteArrayOutputStream().also { bytes ->
            DataOutputStream(bytes).use { out ->
                out.write(CAP_DOMAIN)
                out.write(nonce)
                out.writeShort(version)
                out.write(instanceId.lowercase().toByteArray(Charsets.US_ASCII))
                out.writeShort(name.size)
                out.write(name)
                out.writeShort(caps.size)
                caps.forEach { capability ->
                    out.writeShort(capability.size)
                    out.write(capability)
                }
            }
        }.toByteArray()
    }

    fun transferMessage(nonce: ByteArray, header: WHPP.Header, offset: Long): ByteArray {
        require(nonce.size == 32)
        return ByteArrayOutputStream().also { bytes ->
            DataOutputStream(bytes).use { out ->
                out.write(TRANSFER_DOMAIN)
                out.write(nonce)
                writeTransferFields(out, header, offset)
            }
        }.toByteArray()
    }

    fun receiverMessage(nonce: ByteArray, header: WHPP.Header, offset: Long,
                        receiverInstanceId: String): ByteArray {
        require(nonce.size == 32)
        val receiver = receiverInstanceId.lowercase().toByteArray(Charsets.US_ASCII)
        require(receiver.size == 32)
        return ByteArrayOutputStream().also { bytes ->
            DataOutputStream(bytes).use { out ->
                out.write(RECEIVER_DOMAIN)
                out.write(nonce)
                out.write(receiver)
                writeTransferFields(out, header, offset)
            }
        }.toByteArray()
    }

    private fun writeTransferFields(out: DataOutputStream, header: WHPP.Header, offset: Long) {
        val kind = header.kind.toByteArray(Charsets.UTF_8)
        val filename = header.filename.toByteArray(Charsets.UTF_8)
        require(kind.size <= 0xffff)
        out.write(header.senderInstanceId.lowercase().toByteArray(Charsets.US_ASCII))
        out.write(header.transferId.lowercase().toByteArray(Charsets.US_ASCII))
        out.write(header.sha256.lowercase().toByteArray(Charsets.US_ASCII))
        out.writeShort(kind.size)
        out.write(kind)
        out.writeInt(filename.size)
        out.write(filename)
        out.writeLong(header.plainSize)
        out.writeLong(header.modifiedMs)
        out.writeByte(if (header.encrypted) 1 else 0)
        out.writeLong(offset)
    }

    fun fingerprint(encoded: String): String {
        val raw = Base64.decode(encoded, Base64.DEFAULT)
        decodePublicKey(raw)
        return hex(MessageDigest.getInstance("SHA-256").digest(raw))
    }

    fun verify(encodedPublicKey: String, message: ByteArray, encodedSignature: String): Boolean =
        try {
            val publicKey = decodePublicKey(Base64.decode(encodedPublicKey, Base64.DEFAULT))
            Signature.getInstance("SHA256withECDSA").run {
                initVerify(publicKey)
                update(message)
                verify(Base64.decode(encodedSignature, Base64.DEFAULT))
            }
        } catch (_: Exception) {
            false
        }

    internal fun encodePublicKey(key: ECPublicKey): ByteArray = byteArrayOf(4) +
        fixedCoordinate(key.w.affineX) + fixedCoordinate(key.w.affineY)

    private fun fixedCoordinate(value: BigInteger): ByteArray {
        val raw = value.toByteArray()
        if (raw.size == 32) return raw
        if (raw.size == 33 && raw[0] == 0.toByte()) return raw.copyOfRange(1, 33)
        require(raw.size < 32)
        return ByteArray(32 - raw.size) + raw
    }

    private fun decodePublicKey(raw: ByteArray): ECPublicKey {
        require(raw.size == 65 && raw[0] == 4.toByte())
        val parameters = AlgorithmParameters.getInstance("EC").apply {
            init(ECGenParameterSpec("secp256r1"))
        }.getParameterSpec(ECParameterSpec::class.java)
        val point = ECPoint(
            BigInteger(1, raw.copyOfRange(1, 33)),
            BigInteger(1, raw.copyOfRange(33, 65)),
        )
        return KeyFactory.getInstance("EC").generatePublic(
            ECPublicKeySpec(point, parameters)) as ECPublicKey
    }

    internal fun hex(bytes: ByteArray): String =
        bytes.joinToString("") { "%02x".format(it) }
}

File: src/inkhole/device_identity.py
"""Persistent device identity primitives shared by LAN discovery and WHPP."""

from __future__ import annotations

import base64
import hashlib
import struct

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec


_CAP_DOMAIN = b"INKHOLE-WHPC3\0"
_TRANSFER_DOMAIN = b"INKHOLE-WHPP3-AUTH\0"
_RECEIVER_DOMAIN = b"INKHOLE-WHPP3-RECEIVER\0"


def _encoded_text(value: str, limit: int) -> bytes:
    encoded = value.encode("utf-8")
    if len(encoded) > limit:
        raise ValueError("identity field is too long")
    return encoded


def capability_message(nonce: bytes, instance_id: str, peer_name: str,
                       version: int, capabilities) -> bytes:
    if len(nonce) != 32:
        raise ValueError("capability nonce must be 32 bytes")
    if isinstance(version, bool) or not 0 <= int(version) <= 0xFFFF:
        raise ValueError("capability version is invalid")
    name = _encoded_text(peer_name, 0xFFFF)
    caps = sorted({_encoded_text(str(value), 0xFFFF) for value in capabilities})
    if len(caps) > 0xFFFF:
        raise ValueError("too many capabilities")
    encoded_caps = b"".join(struct.pack("!H", len(value)) + value
                            for value in caps)
    return b"".join((
        _CAP_DOMAIN, nonce, struct.pack("!H", int(version)),
        instance_id.lower().encode("ascii"), struct.pack("!H", len(name)), name,
        struct.pack("!H", len(caps)), encoded_caps,
    ))


def transfer_message(nonce: bytes, header: dict, offset: int) -> bytes:
    if len(nonce) != 32:
        raise ValueError("transfer nonce must be 32 bytes")
    filename = _encoded_text(str(header["filename"]), 0xFFFFFFFF)
    kind = _encoded_text(str(header["kind"]), 0xFFFF)
    return b"".join((
        _TRANSFER_DOMAIN,
        nonce,
        str(header["sender_instance_id"]).lower().encode("ascii"),
        str(header["transfer_id"]).lower().encode("ascii"),
        str(header["sha256"]).lower().encode("ascii"),
        struct.pack("!H", len(kind)), kind,
        struct.pack("!I", len(filename)), filename,
        struct.pack("!QQBQ", int(header["plain_size"]),
                    int(header.get("mtime_ms", 0)),
                    1 if bool(header["encrypted"]) else 0, int(offset)),
    ))


def receiver_message(nonce: bytes, header: dict, offset: int,
                     receiver_instance_id: str) -> bytes:
    """Bind a WHPP resume/ACK channel to the selected receiving device."""
    if len(nonce) != 32:
        raise ValueError("receiver nonce must be 32 bytes")
    receiver = str(receiver_instance_id).lower().encode("ascii")
    if len(receiver) != 32:
        raise ValueError("receiver instance id must be 32 bytes")
    sender_fields = transfer_message(nonce, header, offset)[
        len(_TRANSFER_DOMAIN) + len(nonce):]
    return _RECEIVER_DOMAIN + nonce + receiver + sender_fields


class DeviceIdentity:
    """ECDSA P-256 identity serialized as PKCS#8 for credential-store use."""

    def __init__(self, private_key: ec.EllipticCurvePrivateKey):
        if not isinstance(private_key.curve, ec.SECP256R1):
            raise ValueError("device identity must use P-256")
        self._private_key = private_key
        self.public_bytes = private_key.public_key().public_bytes(
            serialization.Encoding.X962,
            serialization.PublicFormat.UncompressedPoint,
        )
        self.public_key = base64.b64encode(self.public_bytes).decode("ascii")
        self.fingerprint = hashlib.sha256(self.public_bytes).hexdigest()

    @classmethod
    def generate(cls) -> "DeviceIdentity":
        return cls(ec.generate_private_key(ec.SECP256R1()))

    @classmethod
    def from_private_key(cls, encoded: str) -> "DeviceIdentity":
        raw = base64.b64decode(str(encoded), validate=True)
        key = serialization.load_der_private_key(raw, password=None)
        if not isinstance(key, ec.EllipticCurvePrivateKey):
            raise ValueError("device identity is not an EC private key")
        return cls(key)

    def export_private_key(self) -> str:
        raw = self._private_key.private_bytes(
            serialization.Encoding.DER,
            serialization.PrivateFormat.PKCS8,
            serialization.NoEncryption(),
        )
        return base64.b64encode(raw).decode("ascii")

    def sign(self, message: bytes) -> str:
        signature = self._private_key.sign(message, ec.ECDSA(hashes.SHA256()))
        return base64.b64encode(signature).decode("ascii")


def public_fingerprint(encoded: str) -> str:
    raw = base64.b64decode(str(encoded), validate=True)
    if len(raw) != 65 or raw[0] != 4:
        raise ValueError("device public key is not an uncompressed P-256 point")
    ec.EllipticCurvePublicKey.from_encoded_point(ec.SECP256R1(), raw)
    return hashlib.sha256(raw).hexdigest()


def verify(encoded_public_key: str, message: bytes, encoded_signature: str) -> bool:
    try:
        raw_key = base64.b64decode(str(encoded_public_key), validate=True)
        signature = base64.b64decode(str(encoded_signature), validate=True)
        public_key = ec.EllipticCurvePublicKey.from_encoded_point(
            ec.SECP256R1(), raw_key)
        public_key.verify(signature, message, ec.ECDSA(hashes.SHA256()))
        return True
    except (ValueError, TypeError, InvalidSignature):
        return False
